### DISCUSSION TITLE
Creating Test-Users

### DISCUSSION THREAD

**Discussion by lazo23** [imixs/imixs-workflow#793] [Q&A] [answered]
Hi Ralph,

So i have installed imixs-bpmn and created some workflows. Although i want to test the created workflow modells in imixs-office-workflow, but to do that i need some test users. My problem is, that imixs-office-workflow doesn't allow me to create a user without giving an e-mail adress. Can i somehow create users without giving an e-mail adress?
![testuser](https://user-images.githubusercontent.com/103078441/180175841-40f9c1ae-7f27-4eae-b7c0-7246b59e6bd0.PNG)

Can you please help me?

**Comment by rsoika** [comment]
Hi @lazo23, 
for the application [Imixs-Office-Workflow](https://www.office-workflow.de/) indeed the E-Mail field is mandatory. It is possible to overwrite this form for customization but than you also reduce the functionality. So this is not what I would recommand for first tests. 

For testing you can just fill out the field with a dummy address "dummy1@foo.com" .
You can also set the imixs.property

    security.email.unique=false

to allow the same email address for multiple accounts (Administration -> Konfiguration -> weitere Einstellungen)

In addition you can set

    mail.testRecipients=me@foo.com

this setting will catch all outgoing emails and redirect it to a custom list of e-mail addresses. 

In addition there are several ways how to authenticate and manage users within Imixs-Office-Workflow. For example you can also connect the system with a LDAP directory or a SSO module. But even than an E-Mail would be expected. 

Please let me know if this helps you to continue your tests.

**Comment by lazo23** [comment]
Hi @rsoika,

thank you for the quick answer.

I used the a dummy adress as you recommended. This is for this moment enough. Right now, i want to test the ACLs that i defined for the tasks, so it is about to see if the tasks "jump" to the right role. By the way i am modelling a workflow for "permission change". The main idea is to use imixs-office-workflow as a ticket system and i wish to create  single workflows for every relevant ticket within the complany, such as "permission change". 

I also have some problems while modelling a workflows. For example configurating a task, but i will come to you later for that.

**Comment by lazo23** [comment]
Hi @rsoika,

I need help configurating a task for a workflow. How do i define the inputs when the workflow is uploaded to imixs-office-worflow? Can i only do that from Task Properties->Workflow and then edit summary?

For example:

![image](https://user-images.githubusercontent.com/103078441/180401058-627f9f90-d8dd-418f-95ba-2763e71edee4.png)

Is there another way to do that?

Also, can i configurate a workflow with 4 roles? In the given examles i can only see that there are everytime 3 roles. Can i add a second space manager? The reason i want that is, that for a complicated workflow like "permission change" are more than 3 roles involved.

**Reply by rsoika** [reply]
Hi @lazo23 
In your screen you edit the Workflow Summary which is used to display a workitem in a portlet. There you can reference the value of any Item you have defined in your workflow model or in a custom form. There is also a feature which allows you to format the value. See [here](https://www.imixs.org/doc/engine/adapttext.html)

And yes, it is possible to add additional roles to the Organisation diagram. To do this you finally need to create a custom build of Imixs-Office-Workflow and add a JSF/HTML subform named "`sub_space_custom.xhtml`" placed into the folder /src/main/webapp/pages/admin/

The content can look like this:

```html
<ui:composition xmlns="http://www.w3.org/1999/xhtml"
	xmlns:f="http://java.sun.com/jsf/core"
	xmlns:h="http://java.sun.com/jsf/html"
	xmlns:c="http://java.sun.com/jsp/jstl/core"
	xmlns:i="http://xmlns.jcp.org/jsf/composite/imixs"
	xmlns:marty="http://java.sun.com/jsf/composite/marty"
	xmlns:ui="http://java.sun.com/jsf/facelets">
	<div class="imixs-form-panel">
		<div class="imixs-form-section-3">
			<dl>
				<dt>My Extra Role</dt>
				<dd>
					<marty:userListInput value="#{workitem.itemList['mycustom.role']}" editmode="true" />
				</dd>
			</dl>
		</div>
	</div>
</ui:composition>
```
In the worklfow modeler you can add a new entry in the actor-properties:

![image](https://user-images.githubusercontent.com/1339124/180464820-bcf39701-2fb9-4103-8c08-ddf1c02e6ab4.png)


For the data you need to provide a custom plugin, or you extend the TeamPlugin. The goal of the plugin is to provide the new field in your workitem.

**Reply by lazo23** [reply]
Hello @rsoika,

I honestly have some troubles by creating my own workflows model using the Imixs-BPMN. Is there any detailed documentation that could maybe help me?

If there is one, please let me now.

Greetings,
Lazo

**Reply by rsoika** [reply]
Have you take a look [here](https://www.imixs.org/doc/modelling/process.html)?
What exactly is causing you difficulties? Documentation can always be improved ;-)

**Comment by lazo23** [comment]
Yes i have.

So as i have already said, i am trying to customize workflow models for some business processes as "permission change management". So far i have created this workflow model:

![image](https://user-images.githubusercontent.com/103078441/181903786-84a3db6f-1812-459e-8232-19a2dcfa9e94.png)

For every business process are information needed. The information gathering need to be done in the first task of this workflows (Erstellung).

Every time someone want's to creates a new permission change workflow in imixs-office-workflow needs to share the following information, so the change can be done:

- Where he want's the change to happen? (e.g. in which Software) 
- Which permission does he wish to have
- The reason for the change etc. 

![image](https://user-images.githubusercontent.com/103078441/181904287-ced31390-0b80-4cb1-a973-37ebd0f1fb81.png)

This is an example of imixs sultions. How can i custimize my own text fields? Is it something that can be done in imixs-BPMN Modeler?

**Reply by rsoika** [reply]
Hi @lazo23 - yes this information is a little be hidden.
You can find how to work with custom fields [here](https://github.com/imixs/imixs-documents/blob/master/docs/modeling/CUSTOM_FORMS.md).

**Reply by lazo23** [reply]
Hello @rsoika,

Thank you for this information, thanks to that i could customize my own inputs for the workflow.

My problem now is that i want to send a task to the space manager, but it doesn't work. Although i have given the second task the owner, read and write access to the space manager. 

Should i create a select box where the user can select,  which space is responsible for the next task so that the system knows which space is needed?

I have found this example:

	<ui:include src="/pages/workitems/parts/txtspaceref.xhtml">
		<ui:param name="label" value="Der Bereich" />
		<ui:param name="required" value="true" />
		<ui:param name="byprocess" value="false"></ui:param>
	</ui:include>

If I paste that to the custom form description would it  work? I am running Imixs-office-Workflow via docker.

**Reply by rsoika** [reply]
No, the only thing you need to do is adding a custom field into your form description:

    <item name="" type="custom" path="spaceref"  label="Der Bereich:" />

This will show a combobox with all spaces and the `org.imixs.marty.plugins.TeamPlugin` will automatically resolve the teamlist
You can find an example workflow here: https://github.com/imixs/imixs-documents/tree/master/workflow/de

[bedarfsmeldung-de-1.0.0.bpmn](https://github.com/imixs/imixs-documents/blob/master/workflow/de/bedarfsmeldung-de-1.0.0.bpmn) is doing it in a simmilar way

**Reply by lazo23** [reply]
Thank you @rsoika. My workflows model runs perfectly now in imixs-office-workflow. All steps go right the way i wanted.

**Comment by lazo23** [comment]
Hello @rsoika,

my problem now is that, in my workflow model are many spaces involved. In order to solve this problem I have linked the roles with the provided positions. So, to use the role ISB I have defined that ISB is the process assistant, although he is a space manager. How can I link for a different task a different space manager/team/assistant?

Here is the organigram:

![image](https://user-images.githubusercontent.com/103078441/183107176-0ae89d10-d692-4727-9bd3-e8c3949e0253.png)

The following actors are defined in the workflow:

![image](https://user-images.githubusercontent.com/103078441/183097757-75462125-742a-40b5-b05e-ac4ca43fc70d.png)

This is the workflow model:
![image](https://user-images.githubusercontent.com/103078441/183106180-a18e0d84-546c-445f-be0b-ea8fd998c1f4.png)

If I add a second space manager/team/assistant in the actor properties list, will they be linked with the right space manager/team/assistant from imixs-office-workflow? I am asking this because i have tried it and it didnt work. How does the linking work?

Also, i want to create a drop down menu with some software names (i know how to do that), that the user can use to select in which software he wishes his access rights to be changed. For every software is a different space manager responsible. How can i link the software (selection) with different space managers/teams/assistants? The goal is that when user give the needed information, in this case the software name, the system then knows which space is for this software responsible and not the opposite.

**Comment by rsoika** [comment]
Hi @lazo23,
yes, also this requirement can be solved ;-) there are different ways how to achieve this:

With the help of the [TeamPlugin](https://github.com/imixs/imixs-marty/blob/master/doc/team/teamplugin.md) - which you are already using - you can assign a workitem to a static Space or process. Just add the following expression into your Event Workflow Result:

    <item name="space">ISB</item>

This will change the current Space.Ref to the new defined space fully controlled by your model. 

Further more there is a internal intercepter mechnism which you can also use in the modeler mapping of actor roles. You can define an Actor-Mapping like this:

    ISB | {process:ISB:team}

This will resolve the corresponding team list. Find more details [here](https://github.com/imixs/imixs-marty/blob/master/doc/team/teaminterceptor.md)

But be careful: I recommand to assign a workflow instance only to one space and one process. This makes the process more transparent in production. There is one team responsible for a specific task in the model. 
With the help of the [TeamWilldcardInterceptor](https://github.com/imixs/imixs-marty/blob/master/doc/team/teamrolewildcardadapter.md) you can extend the read/write/owner access to workflow instance in a more general way if this is needed. This can have pros and cons. If you add a access role like '{space:?:team}' you can be sure that the space team has always access. But this can also lead to weak access control list. 

With the software selection problem, you can use again the teamPlugin in your model to assign a space depending on a item in your process instance:

    <item name="space"><itemvalue>sofwareselection</itemvalue></item>

The "<itemvalue>sofwareselection</itemvalue>" will be resolved and so you can assign the process instance to the corresponding space.

**Reply by lazo23** [reply]
Hello @rsoika,

before adding the expression that you told me. Do I need to add a workitem into the Task Application Workitem Type? 

Like this:

![image](https://user-images.githubusercontent.com/103078441/183133968-5a26b75c-9893-4bb2-888f-d137bb876eb3.png)

Or its not necessary?

**Reply by lazo23** [reply]
I added into every Task Application Workitem Type "workitem" and instead of <item name="space">ISB</item> I added the following expression: <item name="space">Management.ISB</item> because I got an error, that the space "ISB" is not in the model type. Now i dont get any error, but the workflow does't work properly. It doesn't switch roles, it stays in the role "space-manager" for all the other tasks until the end of the workflow. Can you help me here please?

**Reply by rsoika** [reply]
Sorry, but now it becomes hard to follow you.
First of all, you can leave the "Applicaton -> Workitem Type" empty, as 'workitem' is the default document type. So no need to change this.

Where did you get the error? Where did you have added the expression 'Management.ISB' ? Maybe you can post the exact error message you see on the server console?

**Reply by rsoika** [reply]
You have defined 'ISB' as a sub orgunit of 'Management'. In this case the full name of this Space is 

    Management.ISB

You can verify this if you edit the space. The name of Sub-Spaces are always a concatenation of the parent space and the inner name.

**Reply by lazo23** [reply]
I changed the space from 'ISB' to 'Management.ISB' and I dont get any errors. But the workflow doesn't work properly. 
I added the 'Management.ISB' expression here:

![image](https://user-images.githubusercontent.com/103078441/183217464-0aac5d53-c71e-487c-953c-eab0e8220e08.png)

At this Event:

![image](https://user-images.githubusercontent.com/103078441/183217536-5ac9068b-4de9-4056-a038-b1d31aa208cb.png)

When I click 'freigeben' in imixs-office-workflow the workflow switches to the next task, but the space does't change to 'Management.ISB'. It stays stuck in the same space. 

If it helps you understand the problem. Here is also the ACL for the task 'Genehmigung' that normally should the role 'ISB' check:

![image](https://user-images.githubusercontent.com/103078441/183218828-57553f01-1467-4bf4-b8e3-1900f6db0554.png)

Also here the actor properties for this workflow:

![image](https://user-images.githubusercontent.com/103078441/183218990-b34ad079-93c0-4c5a-9594-8317e75f27f8.png)

Should I change the actor properties too?  From 'ISB | namISB' to 'ISB | Management.ISB?

**Reply by rsoika** [reply]
Yes,  `ISB | namISB` is not correct in your case. If you add the new space ref in the workflow result of the event - as you have done correctly - than you can simply define the ACL by the generic actor mappings

* Bereichs Manager|space.manager
* Bereichs Team|space.team
* Bereichs Assistenz|space.assist 

The team plugin will resolve the members automatically. 

So in your ACL settings you should avoid working with fixed names like 'ISB'. The acl can be handled in a more generic way. Within your events just control to which space or process your workitem should be assigned to. (as you have already done)

Think about the problem with the Software. The model does not need to know all software-teams. It just defines that the space.team is responsible. And in your event you can assign the workitem to the right space depending of the information in your form fields. 

But don't get me wrong here: It is totally fine that you define Lanes labelled with 'ISB' or 'Manager'. This is just for documentation and discussion with the product owner. But technical you work more generic.

**Reply by lazo23** [reply]
Hi @rsoika,

thank you for your answer. Now I understand what you mean I was a little confused at the beginning and I mixed up everything 😂.

So now the workflows actor properties look like this:

![image](https://user-images.githubusercontent.com/103078441/183246250-ddce2543-1bf9-4160-9d80-90d71174292e.png)

I thought, that I need to add every single space as manager, team and assist, which is obviously not right. 

And now the ACL for the Task 'Genehmigung' looks like this:

![image](https://user-images.githubusercontent.com/103078441/183246485-4ef5413f-6552-413a-84b4-d4375cccb520.png)

**Comment by lazo23** [comment]
Hello @rsoika,

You said:

_"Think about the problem with the Software. The model does not need to know all software-teams. It just defines that the space.team is responsible. And in your event you can assign the workitem to the right space depending of the information in your form fields."_

So basically, I could define in the Event Result for which software which space responsible is by checking the information that was selected in the field "Anwendung" (Software).

How does the expression look that I need to add in the Event Result to do that? Can you give me an example?

Programmatically speacking I could do this (Pseudocode):

```
if(Anwendung == "software_name1"){
    space == "management.isb";
}
else if(Anwendung == "software_name2"){
    space == "management.busibess_development";
}
```

Can I do something like that in the Event Result?

**Comment by rsoika** [comment]
Try a script rule like this:

```javascript
var result={};
var _anwendung=''; 
if (  ( workitem.get("Anwendung") != null && !''==workitem.get("Anwendung")[0])   ) {
     _anwendung=workitem.get("Anwendung")[0];
   // set space depnding on anwendung
  if(Anwendung == "software_name1"){
    result.anwendungspace = "management.isb";
  } else if(Anwendung == "software_name2"){
    result.anwendungspace = "management.busibess_development";
  }
} 
```
This will compute a new field named 'anwendungspace'

Than you can set the Event Result to : 

     <item name="space"><itemvalue>anwendungspace</itemvalue></item>

Your situation is very complex. So you can also try to implement a Java Plugin for this mapping. Or - that was what I expected - you create in the Organigram a parent node 'Anwendungen' and place for each Anwendung a child-node defining the responsibilities. Than the user simply selects the Space with the corresponding Anwendung and no mapping is needed.

**Reply by lazo23** [reply]
Hello @rsoika,

I have done as you requested, but I am getting an error in Imixs-Office-Workflow after wanting to switch to the second Task (Sachprüfung)

Here is the error:

![image](https://user-images.githubusercontent.com/103078441/183736789-3de8108d-7177-4d37-b06a-7b539e45752b.png)

Is it possible that a made a mistake with the parameters?
Why does it search for 'Otris privacy Guard' as a space and not for `Management.IT-Beratung`?

Here is the rule script from the Event (Weiterleiten):

![image](https://user-images.githubusercontent.com/103078441/183737399-b99b1968-b7b8-4988-bfec-95f0080b71d2.png)

Here is the expression that I added in the Event Workflow Result (Weiterleiten):

![image](https://user-images.githubusercontent.com/103078441/183737611-85bc8582-93d9-4bf6-8c01-2cd7e40ee15d.png)

And here is the selection form of software (in Task 'Erfassung':

![image](https://user-images.githubusercontent.com/103078441/183738037-d9aa1d30-8061-4e50-8a0a-dd2f906e9302.png)

I think we are on the right way. Can you see what I am doing wrong?

**PS:** The reason I'm doing this is that I'm a dual student in Germany and currently in the practical phase. My practical project is about the introduction of a workflow management tool in the company. I've decided to use Imixs-Office-Workflow for that. Part of my project is to model the following business processes with the modeling tool (Imixs-BPMN): 

- authorization change
- firewall activation
-  contract management.

Also part of the project is to write a manual in german language describing how to model a workflow for internal use. 
I am very thankful, for your answers because they really help me come forward. The final workshop is in ca. 4 weeks :).

**Reply by rsoika** [reply]
Hi,
ok, things can be done much easier. You don't need the javaScript part in your case. 
Just define your Select Menu like this:

```
    ...
    <item name"anwendungspace" type="selectOneMenu" required="true" label="Anwendung"
  options="management.it-beratung|Otris privacy Guard;management.backoffice|GoTo-Meeting" />
```
You can see, I split the name of the Space with a | from the Software name the user choose. Now you can simple set the space to 'anwendungsspace'. 

Only in a much more complex szenario I would recommand to use javaScript Rules or a Java Plugin for the mapping. 

So:

 - remove the rule
 - change the select menu item in your form definition

**Reply by lazo23** [reply]
Hello @rsoika,

thank you for your answer, it worked! 😁

**Comment by lazo23** [comment, marked as answer]
Hello @rsoika,

I know that its possible to upload a file in imixs-office-workflow, but can I somehow make it necassary to the user to upload a file in able switch to the next task? Do I have  to define that in the Event Workflow Rule or can it also be done in a custom-form?

**Reply by rsoika** [reply]
Yes you need to define a validation rule for this. 
As we are currently building a new documentation I have added this information there and also added an example to verify the attachmetns:

https://doc.office-workflow.com/forms/validation.html

Let me know if this answers your question.
